### PR TITLE
Python script to query prometheus for CPU and memory data

### PR DESCRIPTION
**Pull request by wanlin31**
This script is used to query the cpu and memory usage from prometheus for a load test, the results are dumped as a json file. . 

There are three major metrics collected.
* testDurationSeconds is the number of seconds that the Prometheus range query was based on.
* cpuSeconds are the total cpu seconds being used for the duration of a test.
* memoryMean are the average memory usage (in Bytes) for the duration of a test.

The data was arrange as:
```
First: role-of-pod (clients, driver, server)
Second: pod names
Third: container names
Forth: metric names (cpuSeconds, memoryMean)
```
Example usage:

```
python3 prometheus.py --url=http://prometheus.prometheus.svc.cluster.local:9090  --pod_type=clients --container_name=main --container_name=sidecar
```
Example output in gpaste/6456705618542592

### CODE REVIEW

File: tools/run_tests/performance/prom.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    def fetch_by_query(self, query: str) -> Any:
+        """Fetch the given query with time range"""
+        resp = requests.get(self.url + "/api/v1/query_range", {
+            "query": query,
+            "start": self.start,
+            "end": self.end,
+            "step": 5
+        })
+
+        if not resp.ok:
```
Comment (gnossen) [inline tools/run_tests/performance/prom.py:48]: Try [`resp.raise_for_status()`](https://docs.python-requests.org/en/latest/api/#requests.Response) here and throughout.

File: tools/run_tests/performance/prom.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            raise Exception(str(resp))
+
+        return resp.json()
+
+    def fetch_cpu_for_pod(self, container_matcher: str, pod_name: str) -> Dict[str, List[float]]:
+        """Fetch the cpu data for each pod and construct the container
+        name to cpu data list Dict"""
+        query = 'irate(container_cpu_usage_seconds_total{job="kubernetes-cadvisor",pod="' + pod_name + '",container=' + \
+            container_matcher + '}[100s])'
+        if self.debug:
```
Comment (gnossen) [inline tools/run_tests/performance/prom.py:59]: Maybe try using [a logger](https://docs.python.org/3/howto/logging.html) with levels set according to a flag?

**Review by gnossen — APPROVED** [review]

**Review by paulosjca — COMMENTED** [review]
I would call this "prometheus.py" and also the class 'Prometheus", "prom" is cryptic.

**Comment by wanlin31** [comment]
@paulosjca @gnossen I have updated the query to get the total cpu seconds used during the test, also attached the test duration seconds in the results. For the memory, the mean was used to present a single test. The updated example data produced: gpaste/6350045474455552
Let me know if anything need to be changed, thank you!

**Comment by gnossen** [comment]
@wanlin31 In the future, please use `git mv old new` when renaming files. This preserves history so that we can see actual changes after a file has been renamed.

**Comment by gnossen** [comment]
Please resolve the [sanity test issues](https://source.cloud.google.com/results/invocations/275578d7-f2e1-4f8f-b41d-1f39ea927a66/log).

**Review by gnossen — APPROVED** [review]

File: tools/run_tests/performance/prometheus.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    ):
+        self.url = url
+        self.start = start
+        self.end = end
+
+    def _fetch_by_query(self, query: str) -> Any:
+        """Fetches the given query with time range."""
+        resp = requests.get(
+            self.url + "/api/v1/query_range",
+            {"query": query, "start": self.start, "end": self.end, "step": 5},
```
Comment (paulosjca) [inline tools/run_tests/performance/prometheus.py:46]: Let's use only single quotes for strings.

File: tools/run_tests/performance/prometheus.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        self,
+        url: str,
+        start: str,
+        end: str,
+    ):
+        self.url = url
+        self.start = start
+        self.end = end
+
+    def _fetch_by_query(self, query: str) -> Any:
```
Comment (paulosjca) [inline tools/run_tests/performance/prometheus.py:42]: Do we know what is the type of resp.json()? I believe something like Dict[str, Any] might be more appropriate.

**Review by paulosjca — COMMENTED** [review]
Make sure that you pass this through yapf_code.sh.

File: tools/run_tests/performance/prometheus.py (hunk 4 above)
Comment (wanlin31) [reply tools/run_tests/performance/prometheus.py:42]: I will try that and update the type.

File: tools/run_tests/performance/prometheus.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+# A script to fetch total cpu seconds and memory data from prometheus.
+# example usage: python3 prometheus.py
+# --url=http://prometheus.prometheus.svc.cluster.local:9090
+# --pod_type=driver --pod_type=clients --container_name=main
+# --container_name=sidecar
+"""Perform prometheus range queries to obtain cpu and memory data.
+
+This module performs range queries through Prometheus API to obtain
+total cpu seconds and memory during a tets run for given container
```
Comment (paulosjca) [inline tools/run_tests/performance/prometheus.py:25]: test -> test

instant memory *usage*  at the query time.

File: tools/run_tests/performance/prometheus.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        )
+        resp.raise_for_status()
+        return resp.json()
+
+    def _fetch_cpu_for_pod(self, container_matcher: str,
+                           pod_name: str) -> Dict[str, List[float]]:
+        """Fetches the cpu data for each pod.
+
+        Fetch total cpu seconds during the time range specified in the Prometheus instance
+        for a pod. After obtain the cpu seconds, the data are trimmed from time serier to
```
Comment (paulosjca) [inline tools/run_tests/performance/prometheus.py:78]: serier -> series. This happens other times.

File: tools/run_tests/performance/prometheus.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+        logging.debug('raw memory data: %s', str(memory_data))
+        memory_container_name_to_data_list = get_data_list_from_timeseries(
+            memory_data)
+
+        return memory_container_name_to_data_list
+
+    def fetch_cpu_and_memory_data(
+            self, container_list: List[str],
+            pod_dict: Dict[str, List[str]]) -> Dict[str, Any]:
+        """Fetch total cpu seconds and memory data for a multiple pods.
```
Comment (paulosjca) [inline tools/run_tests/performance/prometheus.py:121]: "for multiple pods".

File: tools/run_tests/performance/prometheus.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        containers_to_fetch = '~"' + container_list[0]
+        for container in container_list[1:]:
+            containers_to_fetch = containers_to_fetch + '|' + container
+        containers_to_fetch = containers_to_fetch + '"'
+    return containers_to_fetch
+
+
+def get_data_list_from_timeseries(data: Any) -> Dict[str, List[float]]:
+    """Constructs a Dict as keys are the container names and
+    values are a list of data taken from given timeserie data."""
```
Comment (paulosjca) [inline tools/run_tests/performance/prometheus.py:171]: timeserie -> timeseries.

File: tools/run_tests/performance/prometheus.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+    return container_name_to_data_list
+
+
+def compute_total_cpu_seconds(cpu_data_list: List[float]) -> float:
+    """Computes the total cpu seconds by CPUs[end]-CPUs[start]."""
+    return cpu_data_list[len(cpu_data_list) - 1] - cpu_data_list[0]
+
+
+def compute_average_memory_usage(memory_data_list: List[float]) -> float:
+    """Computes the mean and for given list of data."""
```
Comment (paulosjca) [inline tools/run_tests/performance/prometheus.py:196]: "a" given list of data.

File: tools/run_tests/performance/prometheus.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+        pod_type_to_name["driver"].append(pod_names['Driver']['Name'])
+
+    pod_names_to_query = {}
+    for pod_type in pod_types:
+        pod_names_to_query[pod_type] = pod_type_to_name[pod_type]
+    return pod_names_to_query
+
+
+def convert_UTC_to_epoch(utc_timestamp: str) -> str:
+    """Converts a utc timstamp string to epoch time string."""
```
Comment (paulosjca) [inline tools/run_tests/performance/prometheus.py:227]: timstamp -> timestamp.

**Review by paulosjca — APPROVED** [review]

**Pull request merged**